### DISCUSSION TITLE
how to escape backslash

### DISCUSSION THREAD

**Discussion by dongho-jung** [bpftrace/bpftrace#2022] [Q&A] [answered]
this is a test code:
```c
#include <stdio.h>

int main(void)
{
    printf("hello\nworld\n");
    return 0;
}
```

and I compiled it with `gcc -o test test.c`

when I trace it with `sudo bpftrace -e 'u:libc.so.6:printf { print(str(arg0)) }' -c ./test`
it outputs:
```
Attaching 1 probe...
hello
world
```

is there any ways to output as follows?
```
Attaching 1 probe...
hello\nworld\n
```

thanks!

**Comment by mmisono** [comment, marked as answer]
You can escape the backslash with a backslash.

```
% sudo bpftrace -e 'BEGIN { printf("hello\\ world\n");}' 
Attaching 1 probe...
hello\ world
```

**Reply by dongho-jung** [reply]
but as you can see, I'm receiving string as parameter not literal.
I tried escape it outside bpftrace though, it wasn't simple task (if there's straightforward solution, please let me know)
I read [reference_guide](https://github.com/iovisor/bpftrace/blob/master/docs/reference_guide.md), [tutorial_one_liners](https://github.com/iovisor/bpftrace/blob/master/docs/tutorial_one_liners.md) but it seems there's no answer for this.

I think I gotta do that with other shell commands, but I wish I've missed something

**Reply by mmisono** [reply]
Ah, I see. Maybe we should implement some escape function but for now, we can use json output (`-f json`) and query results using `jq`.

```
% sudo ./bpftrace -e 'u:libc.so.6:puts{ print(str(arg0)) }' -c ./test -f json | jq -R 'fromjson? | select (.type == "value") | .data'
"hello\nworld"
```

**Reply by dongho-jung** [reply]
oh wow I've missed `-f FORMAT` option!
however when I tried that I got just empty string..

```
!?master ~/projects/snippets/playground/c/bpftrace> cat test2.c 
#include <stdio.h>

int main(void)
{
    printf("hello\nworld\n");
    return 0;
}

!?master ~/projects/snippets/playground/c/bpftrace> objdump -tT test2
...
0000000000000000      DF *UND*	0000000000000000  GLIBC_2.2.5 puts
...


!?master ~/projects/snippets/playground/c/bpftrace> sudo bpftrace -e 'u:libc.so.6:puts{ print(str(arg0)) }' -c ./test2 -f json
{"type": "attached_probes", "data": {"probes": 1}}
hello
world
{"type": "value", "data": ""}  # why do I get empty data..?
```

puts(3) says its signature is `int puts(const char *s);` so I think arg0 should mean the buffer to be written. But I have no idea why do I get empty data..

and this is not related to only -f json, it also occurs with -f text
... the point is problem seems to be changed. should I open the new issue?

**Reply by mmisono** [reply]
I guess the problem is the same as the #2005. In short, when accessing `arg0` page fault occurs. Since BPF cannot handle page faults, it returns a null string. A workaround is using `uretprobe` in the same way that I presented in #2005.

**Reply by mmisono** [reply]
i.e.,
```
% sudo ./bpftrace  -e 'u:/lib/x86_64-linux-gnu/libc-2.31.so:puts { @[tid]=arg0 } ur:/lib/x86_64-linux-gnu/libc-2.31.so:puts /@[tid]/ {print(str(@[tid])) }' -c ./a.out -f json | jq -R 'fromjson? | select (.type == "value") | .data'
"hello\nworld"
```

**Reply by dongho-jung** [reply]
yes your answer is totally right and it works fine now! thanks a lot!
